src/Compilers/Core/Portable/Serialization/ObjectBinder.cs

```diff
@@ -2,79 +2,107 @@
 
 using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
 
 namespace Roslyn.Utilities
 {
     /// <summary>
-    /// A <see cref="ObjectBinder"/> that records runtime types and object readers during object writing so they
-    /// can be used to read back objects later.
+    /// <see cref="ObjectBinder"/> is a registry that maps between arbitrary <see cref="Type"/>s and 
+    /// the 'reader' function used to deserialize serialized instances of those types.  Registration
+    /// must happen ahead of time using the <see cref="RegisterTypeReader"/> method.
     /// </summary>
-    /// <remarks>
-    /// This binder records runtime types an object readers as a way to avoid needing to describe all serialization types up front
-    /// or using reflection to determine them on demand.
-    /// </remarks>
     internal static class ObjectBinder
     {
+        /// <summary>
+        /// Lock for all data in this type.
+        /// </summary>
         private static object s_gate = new object();
-        private static readonly Dictionary<Type, int> s_typeToIndex = new Dictionary<Type, int>();
 
+        /// <summary>
+        /// A monotonically increasing counter we have that gets changed every time a new deserialization
```

Comment:
"Monotonically" means it can stay the same over time, but this is strictly increasing.